source/extensions/filters/http/rate_limit_quota/filter.cc

```diff
@@ -24,18 +72,34 @@ Http::FilterHeadersStatus RateLimitQuotaFilter::decodeHeaders(Http::RequestHeade
 
   // Second, generate the bucket id for this request based on match action when the request matching
   // succeeds.
-  const RateLimitOnMatchAction* match_action =
-      dynamic_cast<RateLimitOnMatchAction*>(match_result.value().get());
-  auto ret = match_action->generateBucketId(*data_ptr_, factory_context_, visitor_);
+  const RateLimitOnMatchAction& match_action =
+      match_result.value()->getTyped<RateLimitOnMatchAction>();
+  auto ret = match_action.generateBucketId(*data_ptr_, factory_context_, visitor_);
```

Comment:
This is another place where computation will be expensive for each request. I think it will make sense to create mapping from BucketId to uint64_t hash once when they received the first time and then just use the hash for bookkeeping.

Reply:
if I understand your comment correctly, I don't think we can avoid `generateBucketId` on each request . It is because BucketId is generated based on request matching and request attributes passed in here. 
In other words, we need to know for each request if it is mapped to certain bucket or not by generating the bucket based on request matching.

However, I did avoid using bucketId proto as key, which I think should significant reduce the cost as all the map look up will not be performed on proto.

Reply:
Ah, I see. I forgot the the bucket ID can be generated from request properties. So yes, you are correct, it needs to be computed for each request.